Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.h

```diff
@@ -50,42 +51,66 @@ class AudioSampleDataSource;
 
-    // AudioMediaStreamTrackRendererInternalUnit
-    OSStatus render(size_t sampleCount, AudioBufferList&, uint64_t sampleTime, double hostTime, AudioUnitRenderActionFlags&) final;
-    void reset() final;
-
     void retrieveFormatDescription(CompletionHandler<void(std::optional<CAAudioStreamDescription>)>&&);
 
     // BaseAudioMediaStreamTrackRendererUnit
-    void setAudioOutputDevice(const String&) final;
-    void addSource(Ref<AudioSampleDataSource>&&) final;
-    void removeSource(AudioSampleDataSource&) final;
-    void addResetObserver(ResetObserver& observer) final { m_resetObservers.add(observer); }
+    void addSource(const String&, Ref<AudioSampleDataSource>&&) final;
+    void removeSource(const String&, AudioSampleDataSource&) final;
+    void addResetObserver(const String&, ResetObserver&) final;
 
 private:
     AudioMediaStreamTrackRendererUnit();
 
-    void start();
-    void stop();
-
-    void createAudioUnitIfNeeded();
-    void updateRenderSourcesIfNecessary();
-
-    Ref<AudioMediaStreamTrackRendererInternalUnit> protectedInternalUnit() { return m_internalUnit; }
-
-    HashSet<Ref<AudioSampleDataSource>> m_sources;
-    Vector<Ref<AudioSampleDataSource>> m_pendingRenderSources WTF_GUARDED_BY_LOCK(m_pendingRenderSourcesLock);
-    Vector<Ref<AudioSampleDataSource>> m_renderSources;
-    bool m_hasPendingRenderSources WTF_GUARDED_BY_LOCK(m_pendingRenderSourcesLock) { false };
-    Lock m_pendingRenderSourcesLock;
-    Ref<AudioMediaStreamTrackRendererInternalUnit> m_internalUnit;
-    WeakHashSet<ResetObserver> m_resetObservers;
+    void deleteUnitsIfPossible();
+
+    class Unit : public AudioMediaStreamTrackRendererInternalUnit::Client {
+    public:
+        static Ref<Unit> create(const String& deviceID) { return adoptRef(*new Unit(deviceID)); }
+        ~Unit();
+
+        void addSource(Ref<AudioSampleDataSource>&&);
+        bool removeSource(AudioSampleDataSource&);
+        void addResetObserver(ResetObserver&);
+        bool hasSources() const
+        {
+            assertIsMainThread();
+            return !m_sources.isEmpty();
+        }
+
+        bool isDefault() const { return m_isDefaultUnit; }
+
+        void close();
+        void retrieveFormatDescription(CompletionHandler<void(std::optional<CAAudioStreamDescription>)>&&);
+
+    private:
+        explicit Unit(const String&);
+
+        void start();
+        void stop();
+        void updateRenderSourcesIfNecessary();
+
+        // AudioMediaStreamTrackRendererInternalUnit::Client
+        OSStatus render(size_t sampleCount, AudioBufferList&, uint64_t sampleTime, double hostTime, AudioUnitRenderActionFlags&) final;
+        void reset() final;
+
+        HashSet<Ref<AudioSampleDataSource>> m_sources WTF_GUARDED_BY_CAPABILITY(mainThread);
+        Vector<Ref<AudioSampleDataSource>> m_pendingRenderSources WTF_GUARDED_BY_LOCK(m_pendingRenderSourcesLock);
+        Vector<Ref<AudioSampleDataSource>> m_renderSources;
```

Comment:
couldn't this have an GUARDED_BY , either a lock or a given thread.

Reply:
Not directly, since it is only accessed on the rendering thread, which we cannot directly assert.
I'll do that as a follow-up, we can probably get the assertion from the underlying unit.